Source/Orts.Simulation/Simulation/RollingStocks/MSTSWagon.cs

```diff
@@ -1944,15 +1953,34 @@ public override void Update(float elapsedClockSeconds)
                     FreightAnimations.LoadedOne = null;
                     FreightAnimations.FreightType = PickupType.None;
                 }
-                if (FreightAnimations.WagonEmptyWeight != -1) MassKG = FreightAnimations.WagonEmptyWeight + FreightAnimations.FreightWeight + FreightAnimations.StaticFreightWeight;
-                if (WaitForAnimationReady && WeightLoadController.CommandStartTime + FreightAnimations.UnloadingStartDelay <= Simulator.ClockTime)
+                                if (WaitForAnimationReady && WeightLoadController.CommandStartTime + FreightAnimations.UnloadingStartDelay <= Simulator.ClockTime)
                 {
                     WaitForAnimationReady = false;
                     Simulator.Confirmer.Message(ConfirmLevel.Information, Simulator.Catalog.GetString("Starting unload"));
                     if (FreightAnimations.LoadedOne is FreightAnimationContinuous)
                         WeightLoadController.StartDecrease(WeightLoadController.MinimumValue);
                 }
             }
+
+            if (WagonType != WagonTypes.Tender && AuxWagonType != "AuxiliaryTender" && WagonType != WagonTypes.Engine)
+            {
+                // Updates mass when it carries containers
+                float totalContainerMassKG = 0;
+                if (FreightAnimations?.Animations != null)
+                {
+                    foreach (var anim in FreightAnimations.Animations)
+                        if (anim is FreightAnimationDiscrete discreteAnim && discreteAnim.Container != null)
+                        {
+                            totalContainerMassKG += discreteAnim.Container.MassKG;
+                        }
+                }
+
+                // Updates the mass of the wagon considering all types of loads
+                if (FreightAnimations != null && FreightAnimations.WagonEmptyWeight != -1)
+                {
+                    MassKG = FreightAnimations.WagonEmptyWeight + FreightAnimations.FreightWeight + FreightAnimations.StaticFreightWeight + totalContainerMassKG;
+                }
+            }
```

Comment:
Consider whether this code duplication can be removed by creating a CalculateMass() function. Same code here as in LoadFromWagFile.

Reply:
Code duplication removed.